pkg/sessions/redis/client.go

```diff
@@ -18,16 +22,54 @@ var _ Client = (*client)(nil)
 
 type client struct {
 	*redis.Client
+	locker *redislock.Client
+	locks  map[string]*redislock.Lock
 }
 
 func newClient(c *redis.Client) Client {
-	return &client{Client: c}
+	return &client{
+		Client: c,
+		locker: redislock.New(c),
+		locks:  map[string]*redislock.Lock{},
```

Comment:
What's the rationale for tracking locks ourselves vs relying on underlying errors returned by `locker.Obtain` if a lock is already taken?

Do we run the risk of drift if we try to manage assignment & expiration in 2 different spots (in our code, & redis data store)? I understand the benefits of shaving roundtrips to Redis, so something to weight pros and cons of.

Reply:
While this also tracks all locks potentially acquired in a single process space -- this would fall apart in horizonatally scaled deployments right?

Reply:
Based on the redislock implementation, I need to know the lock to release it. I simply call `lock.Release()`. Thats why I'm persisting it here.

Reply:
So I think this might have issues since maps aren't safe for concurrency without using a mutex or channels. Plus it feels weird to manage locks on the client.

What about this design:

We have our own Lock interface (calling out `Obtain` / `Release`, whatever else we need). Then we store the Lock in whatever `SessionState` we pop out of our `SessionStore` (so our store will still `Save`, `Load`, `Clear`).

So the Redis store will attach a `Lock` implementation that is `redislock` aware to the created `SessionState`, Cookie will attach its own noop Lock.

Then we can fully manage and control the lock from the SessionState if they ever have an operation that needs a lock (e.g. refresh) - that feels more natural to me than trying to track all existing locks from the client.

Reply:
This design would go better once we refactored our `RefreshSession` implementation to allow refreshing before the Access Token has expired.

Then 1 request `SessionState` could grab the lock for refreshing, while the others peek (we can check the redis key) and skip refreshing and carry on proxying like normal if the refreshing lock is claimed by another request.

Otherwise, if you can't refresh until the access token is expired, I take it you want to have all requests wait until the session is refreshed? Or can we be greedy and have them be slightly stale and the follow-up requests that aren't concurrent with the one that grabbed the refresh lock will be able to pull up the refreshed session 🤔

Reply:
> We have our own Lock interface (calling out Obtain / Release, whatever else we need). Then we store the Lock in
> whatever SessionState we pop out of our SessionStore (so our store will still Save, Load, Clear).
> 
> So the Redis store will attach a Lock implementation that is redislock aware to the created SessionState, Cookie will 
> attach its own noop Lock.
> 
> Then we can fully manage and control the lock from the SessionState if they ever have an operation that needs a lock > (e.g. refresh) - that feels more natural to me than trying to track all existing locks from the client.

Yes, I like the idea. I will try to change it like that.

I also like the idea of refreshing when the access token is still valid for X amount of time. In this case we can actually just skip refreshing. However, this would increase the necessary changes here in the PR again. What do you suggest how to proceed here?

What do you think about my further adjustments regarding the isRefreshSessionNeededWithProvider method?

Reply:
Let's solely handle locking in this PR -- its a complex topic and it will help @JoelSpeed and I fully understand the nuances of just that as we wrap our brains around it.

Early refreshing is another complex refactor with its share of pitfalls. We are tracking it here: https://github.com/oauth2-proxy/oauth2-proxy/issues/856